webapp/src/main/java/org/mediacloud/cliff/servlet/service/ParseTextService.java

```diff
@@ -0,0 +1,166 @@
+                .map(a -> (Long) a.getResult().get("milliseconds"))
+                .map(a -> a == null ? 0 : a)
+                .reduce(Long::sum)
+                .orElse(0L);
+        executorService.shutdown();
+        return BatchRawParseTextResponse.builder()
+                .results(responses)
+                .actualMilliseconds(stopWatch.getTime())
+                .sumOfMilliseconds(milliseconds)
+                .build();
+    }
+
+    private static Map<Pair<String, String>, Long> toCountryStateCountMap(List<Mention> mentions) {
+        return mentions.stream()
+                .map(m -> Pair.of(m.getCountryCode(), m.getStateCode()))
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+    }
+
+    private static Map<String, Long> toCountryMap(List<Mention> mentions) {
+        return mentions.stream()
+                .map(Mention::getCountryCode)
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+    }
+
+    private static Optional<String> getCountyName(Mention mention) {
+        HashMap<?, ?> geonameInfo = (HashMap<?, ?>) ParseManager.getGeoNameInfo(mention.getId()).get("results");
+
+        while (geonameInfo.containsKey("parent")) {
+            if (COUNTY_FEATURE_CODE.equals(geonameInfo.get("featureCode"))) {
+                return Optional.of((String) geonameInfo.get("name"));
+            }
+            geonameInfo = (HashMap<?, ?>) geonameInfo.get("parent");
+        }
+        
+        return Optional.empty();
+    }
+
+    private static ParseTextResponse collapseMentions(ParseTextResponse parseTextResponse) {
+        Map<Pair<String, String>, Long> countryStateCountMap = toCountryStateCountMap(parseTextResponse.getMentions());
+        Map<String, Long> countryCountMap = toCountryMap(parseTextResponse.getMentions());
+
+        // Getting most specific place grouping by country and state
+        List<Mention> condensedMentions = parseTextResponse.getMentions().stream()
+                // Getting most specific place grouping by country and state
+                .filter(mention -> {
+                    if (COUNTRY_FEATURE_CODE.equals(mention.getFeatureCode()) && !countryCountMap.get(mention.getCountryCode()).equals(1L)) {
+                        return false;
+                    }
+
+                    Pair<String, String> countryStateKey = Pair.of(mention.getCountryCode(), mention.getStateCode());
+                    return !STATE_FEATURE_CODE.equals(mention.getFeatureCode()) || countryStateCountMap.get(countryStateKey).equals(1L);
+                })
+                // counties need to have the "county" field populated
+                .peek(mention -> {
+                    if (COUNTY_FEATURE_CODE.equals(mention.getFeatureCode())) {
+                        mention.setCounty(mention.getPlace());
+                    }
+                })
+                // "places" need the county field populated.
```

Comment:
Are we including cities (PPL) as places too? In my notes in the document I have that cities won't be filling up the county, but if it doesn't affect performance we could leave it

Reply:
We are not, but I'd say we can. Let's run some tests first in CI to make the final decision

Reply:
Got it, but we're still taking cities into account for grouping the following case?

US,_,_
US, FL, _
US, FL, Miami

Only returns
US, FL, Miami

Reply:
I just did a quick test on my side with:

```
I'm from United States. I live in the Florida state, specifically in gainesville
```

... and I'm getting only:

US,FL,_,Gainesville

(the _ is for the county)

Reply:
Perfect